File: vsh/switch_connector.py
from typing import Literal

from amaranth import Signal

import sim


class SwitchConnector:
    switch: Signal

    _state: Literal["idle", "pressing", "releasing"]

    def __init__(self, switch: Signal):
        self.switch = switch

        self._state = "idle"

    def sim_process(self) -> sim.Generator:
        while True:
            if self._state == "pressing":
                self._state = "releasing"
                yield self.switch.eq(1)
            elif self._state == "releasing":
                self._state = "idle"
                yield self.switch.eq(0)
            yield

    def press(self):
        assert self._state == "idle"
        self._state = "pressing"

File: vsh/src/SwitchConnector.zig
const Cxxrtl = @import("./Cxxrtl.zig");

object: Cxxrtl.Object(bool),
state: enum { Idle, Pressing, Releasing } = .Idle,

pub fn init(swi: Cxxrtl.Object(bool)) @This() {
    return .{
        .object = swi,
    };
}

pub fn tick(self: *@This()) void {
    switch (self.state) {
        .Idle => {},
        .Pressing => {
            self.state = .Releasing;
            self.object.next(true);
        },
        .Releasing => {
            self.state = .Idle;
            self.object.next(false);
        },
    }
}

pub fn press(self: *@This()) void {
    if (self.state == .Idle) {
        self.state = .Pressing;
    }
}
